core/src/main/java/org/keycloak/representations/idm/ClientTypeRepresentation.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.keycloak.representations.idm;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class ClientTypeRepresentation {
```

Comment:
Since this is primarily a data holding class, I wonder if it would be a good candidate to use [Java records](https://docs.oracle.com/en/java/javase/17/language/records.html) instead?

It could look something like...

```
public record ClientTypeRepresentation(
        @JsonProperty("name")
        String name,
        @JsonProperty("provider")
        String provider,
        @JsonProperty("config")
        Map<String, PropertyConfig> config,
        @JsonProperty("referenced-properties")
        Map<String, Object> referencedProperties
) {

    public ClientTypeRepresentation {
        if (referencedProperties == null) {
            referencedProperties = new HashMap<>();
        }
    }

    public record PropertyConfig(
            @JsonProperty("applicable")
            Boolean applicable,
            @JsonProperty("read-only")
            Boolean readOnly,
            @JsonProperty("default-value")
            Object defaultValue
    ) { }
}
```

Reply:
Though I don't know if there are any constrains placed on this by the admin client - I'm assuming its expected to be run with Java 17.

Reply:
I'm going to try changing this to a record in my next MR, since most of the logic is there and it'll be easier to tell if it breaks something.

Reply:
AFAIK keycloak-admin-client is still expected to run with older Java versions. That is why we don't use this stuff on other representation classes. So my vote is to rather keep it similar like other representation classes inside package `org.keycloak.representations.idm`

Reply:
Ah, that makes sense then. Thanks for the insight @mposolda !